### PR TITLE
[query] Add Graphite support for ** with metric path selectors

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

Adds support for `**` for Graphite path selectors.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/query/graphite/storage/m3_wrapper.go [hunk 1]
```diff
@@ -94,6 +96,32 @@ func NewM3WrappedStorage(
...
+		globOpts := graphite.GlobOptions{
+			AllowMatchAll: true,
+		}
+		idRegexp, _, err := graphite.ExtendedGlobToRegexPattern(query, globOpts)
+		if err != nil {
+			return nil, err
+		}
+		// Also make sure it's a graphite metric by making sure it has a
+		// __g0__ tag:
+		hasFirstPathMatcher, err := convertMetricPartToMatcher(0, wildcard)
```
Comment (yyinsc) [inline src/query/graphite/storage/m3_wrapper.go:111]: this check seems can happen first, before the glob handling?

File: src/query/graphite/storage/m3_wrapper.go [hunk 2]
```diff
@@ -94,6 +96,32 @@ func NewM3WrappedStorage(
...
+		if err != nil {
+			return nil, err
+		}
+		return models.Matchers{
+			hasFirstPathMatcher,
+			models.Matcher{
+				Type:  models.MatchRegexp,
+				Name:  doc.IDReservedFieldName,
```
Comment (martin-mao) [inline src/query/graphite/storage/m3_wrapper.go:119]: Will looking in the ID field for this fail for search patterns with a "." in it?

Comment (robskillington) [reply src/query/graphite/storage/m3_wrapper.go:119]: `ExtendedGlobToRegexPattern` which we call here takes care of this:
```go
		case '.':
			// Match hierarchy separator
			pattern.WriteString("\\.+")
			regexed = true
```

i.e. it escapes the dot and allows for searching patterns with dots in it.

File: src/query/graphite/storage/m3_wrapper.go (hunk 1 above)
Comment (robskillington) [reply src/query/graphite/storage/m3_wrapper.go:111]: Sure thing.

**Review by martin-mao — APPROVED** [review]
LGTM

**Pull request merged**